arangod/Cluster/FollowerInfo.cpp

```diff
@@ -162,28 +162,39 @@ Result FollowerInfo::remove(ServerID const& sid) {
                                          // local data is modified again.
 
   // First check if there is anything to do:
-  if (std::find(_followers->begin(), _followers->end(), sid) == _followers->end()) {
-    TRI_ASSERT(std::find(_failoverCandidates->begin(), _failoverCandidates->end(),
-                         sid) == _failoverCandidates->end());
+  if (std::find(_followers->begin(), _followers->end(), sid)
+        == _followers->end() &&
+      std::find(_failoverCandidates->begin(), _failoverCandidates->end(), sid)
+        == _failoverCandidates->end()) {
     return {TRI_ERROR_NO_ERROR};  // nothing to do
   }
-  // Both lists have to be in sync at any time!
+  // Both lists have to be in sync most of the time, sometimes the
+  // _failoverCandidates have additional servers, and sometimes, this
+  // code here is called to remove a server from the _failoverCandidates,
+  // even if it is not on _followers. There should never be a follower
+  // which is in _followers but not in _failoverCandidates. Therefore,
+  // there should never be a call to remove here for a server which is
+  // not in the failoverCandidates.
+  // Note that it is possible that _followers is empty, when we have to
+  // remove a failoverCandidate.
   TRI_ASSERT(std::find(_failoverCandidates->begin(), _failoverCandidates->end(),
                        sid) != _failoverCandidates->end());
   auto oldFollowers = _followers;
   auto oldFailovers = _failoverCandidates;
   {
     auto v = std::make_shared<std::vector<ServerID>>();
-    TRI_ASSERT(!_followers->empty());  // well we found the element above \o/
-    v->reserve(_followers->size() - 1);
+    if (!_followers->empty()) {
+      v->reserve(_followers->size() - 1);
+    }
```

Comment:
I vote for removing the if ... reserve altogether. It is a micro optimization that adds complexity and already has led to a problem before. Not reserving space will not kill us performance-wise, as we are talking about 1 or 2 followers only.

Reply:
I agree the resizing here does not really help, and can be removed.